File: 3rd Semester/Advanced Programming Methods(Done)/Laboratory/src (BUX)/Model/Statements/FileOperations/openRFileStmt.java
package Model.Statements.FileOperations;

import Model.Statements.IStmt;
import Model.Utils.MyIDictionary;
import Model.Utils.IFileTable;
import Model.PrgState;
import Exceptions.MyControllerException;
import Model.Utils.Tuple;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;

public class openRFileStmt implements IStmt
{
    private String varName;
    private String filename;

    public openRFileStmt(String varName, String filename )
    {
        this.varName = varName;
        this.filename = filename;
    }

    public PrgState execute(PrgState state) throws MyControllerException {
        MyIDictionary<String, Integer> symtab = state.getSymTable();
        IFileTable<Tuple> fileT = state.getFT();

        try{
            for(int i = 1; i<=fileT.getKey();i++)
            {
                Tuple tpl = fileT.get(i);
                if(tpl != null && filename.equals(tpl.getName()))
                    throw new MyControllerException("The file already exists!");
            }
            BufferedReader bf = new BufferedReader(new FileReader(this.filename));
            Tuple tpl = new Tuple(filename, bf);
            fileT.add(tpl);
            symtab.put(this.varName, fileT.getKey());
        }catch (FileNotFoundException e)
        {
            System.out.println(e.toString());
        }
        return null;
    }
    public String toString() {
        return "Opening: " + filename;
    }
}

File: 3rd Semester/Advanced Programming Methods/Laboratory/Toy Interpreter C#/sharplab/Model/Statements/openRFileStmt.cs
using sharplab.Model.Utils;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace sharplab.Model.Statements
{
    class OpenRFileStmt : IStmt
    {
        private String varName;
        private String filename;

        public OpenRFileStmt(string varName, string filename)
        {
            this.varName = varName;
            this.filename = filename;
        }

        public PrgState Execute(PrgState state)
        {
            IMyDictionary<string, int> symtab = state.GetSymTable();
            IFileTable<Utils.Tuple> fileT = state.GetFT();

            try
            {
                for(int i = 1; i<=fileT.GetKey();i++)
                {
                    Utils.Tuple tupl = fileT.Get(i);
                    if(tupl != null && filename.Equals(tupl.GetName()))
                        throw new Exception("The file already exists!");
            }
            StreamReader s = new StreamReader(this.filename);
            Utils.Tuple tpl = new Utils.Tuple(filename, s);
            fileT.Add(tpl);
            symtab.Put(this.varName, fileT.GetKey());
            }
            catch (FileNotFoundException e)
            {
                System.Console.WriteLine(e.ToString());
            }
            return state;
        }
        public override string ToString()
        {
            return "Opening: " + filename;
        }
    }
    
}
